### ISSUE TITLE
Variables concept map

### ISSUE THREAD

**Issue by tischi** [NEUBIAS/training-resources#369]
@manerotoni 

<img width="837" alt="image" src="https://user-images.githubusercontent.com/2157566/159178274-34337b9b-c424-4fcd-8415-b5f4e069146b.png">

I am not sure, it may be semantics, but I would rather teach that typically the type of a variable does _not_ change. 

If you think of such an example:
```
a = 1;
a = "hello";
```
Yes, the type is changing but I would rather say that in the second line of code this is a _new_ variable with a new address in memory such that the type has not changed; and you cannot access the first variable anymore because you overloaded the symbolic name.

Again, not sure, pinging some experts:

@unode how you are you teaching the concept of a variable?
@constantinpape what do you think?

**Comment by unode** [comment]
Depends at what level. For beginners I usually leave it at "a name" for a value.

Things like type-bound variables are only relevant in some languages of which Python isn't really one.
Teaching that the type doesn't change can lead to confusion because the compiler will not complain if the type of a variable changes.
You will only get warnings if you use something like `mypy` to type-check the code which is a topic for intermediate-advanced level (type annotations).

"Stored in memory" can be misleading. The object is the thing that is stored in memory, variables are just names/alias for humans to read. The compiler doesn't care about names.

**Comment by tischi** [comment]
> Teaching that the type doesn't change can lead to confusion because the compiler will not complain if the type of a variable changes.

I think it depends what one means by "variable".
It could be:
1. The thing for which there is space allocated in memory
2. The name that you use to address this thing

I would say for (1) the type cannot change and for (2) it can change. 

Maybe we should simplify the concept map to avoid this discussion?

**Comment by constantinpape** [comment]
My 2cents:
- variable types in python can and do change all the time (e.g. one of the most common python pattern that every beginner should learn is to have default value `None` for variables; this is a different type than what the variable will contain; and that's just the most basic example). So if this should at all be relevant for python (or other dynamically typed languages) then `types` can change. (But this statement is of course not true for statically typed languages).
- memory: I would not talk at all about memory when introducing the concept of a variable. The mapping of variables to memory location depends a lot on the language and then even in the language about the object a variable is holding (e.g. stack vs heap). On a trivial level, `stored in memory` is always true, so why mention it at all.

>  1. The thing for which there is space allocated in memory
>   2. The name that you use to address this thing
> 
> I would say for (1) the type cannot change and for (2) it can change.
> 
> Maybe we should simplify the concept map to avoid this discussion?

Neither 1 or 2 are very good definition of variables because focus on memory is the wrong level of abstraction for variables.

**Comment by unode** [comment]
Sounds like even just now we used different interpretations. Indeed for me variable usually refers to the name but reading on it online it seems I'm on the wrong.

In my own mental model, I use only `name` and `object`. The latter stored in memory and containing a type. However with the newer type-annotations feature, I now understand this as names now having types too.

To me `variable` is a synonym of `name`.

**Comment by manerotoni** [comment]
Depending on the language, 
the type of a variable is allowed to change and this is what I wanted to teach. It may be a little complicated but it is important to know that this can cause confusions, errors or completely mess up your code.

**Comment by tischi** [comment]
Thanks for all the input!
Looks like, in fact, my internal concept map of a variable is/was wrong :-)

> To me variable is a synonym of name.

If that's the definition, then of course the type can change. But then, I personally would consider changing our concept map, because then a variable does not _have_ a name, but it _is_ the name.

**Comment by constantinpape** [comment]
> If that's the definition, then of course the type can change. But then, I personally would consider changing our concept map, because then a variable does not _have_ a name, but it _is_ the name.

I don't think that this is really correct either, the variable is not directly the name, because the name is some string of symbols that represent the variable, like `a`. But the variable still have value and type (which, as discussed, can change).
The name is somehow closer associated with the variable because it cannot change, but there's obviously more to a variable than its name.

To me the concept map looks pretty good as it is. Besides maybe the `is stored in memory`, because this is a) trivially true, everything that runs inside of a computer program is somehow stored in memory. b) how exactly this is done depends on the language, type, etc. (But to be clear I don't think that putting the `stored in memory` is wrong.)

**Comment by tischi** [comment]
What do you think about something like this as a concept map?

`name --> object`
`object ---|has| type`
`object ---|has| value`

In that sense the word variable would not appear in the concept map, but the concept of a variable would be the mapping of a name to an object.

**Comment by tischi** [comment]
I'll reopen, such that after the course we could give this a second thought.
